### ISSUE TITLE
.onceover directory sync problems

### ISSUE THREAD

**Issue by GeoffWilliams** [dylanratcliffe/onceover#113]
Sometimes get strange problems synchronising current control repo directory with the one in `.onceover` which lead to stale files being used for tests.  This definitely occurs when `--skip_r10k` is used and results in internal roles and profiles not being updated.  I had to resort to workarounds like this:

```
		rm .onceover/etc/puppetlabs/code/environments/production/site/ -rf ; \
		rm .onceover/etc/puppetlabs/code/environments/production/hieradata -rf ; \
		cp site .onceover/etc/puppetlabs/code/environments/production/ -r ; \
		cp hieradata .onceover/etc/puppetlabs/code/environments/production/ -r ; \
```
In order to get non-stale data.
Without --skip-r10k, sometimes modules don't update occasionally and the fix is `rm -rf .onceover`

Lets collect the list of problems here so they can be hunted down

**Comment by GeoffWilliams** [comment]
Duplicate of https://github.com/dylanratcliffe/onceover/issues/95 I think

I'm getting really insidious behaviour from this - all my hiera lookups suddenly stopped working due to a stale hiera.yaml in my .onceover directory that must not have been updated in days suddenly updating to a wrong file that I thought was already being used  - its making my head hurt :rage: :confounded: 

Haven't looked at the code yet but what do you think about this as an approach:
1. Configure r10k to use a global cache dir outside of `.onceover` - perhaps at `~/.onceover` or `~/.puppetlabs` - by externalising the cache it can be shared between customers/control repos your working with - could also be used for the benefit of other tools too if we pick a common location
2.  Destroy `.onceover` at the start of the testing process
3.  Copy the control repo files into the place they normally go
4.  If using R10K, do an `r10k puppetfile install` and install into a new directory at the top level of the control repo - something like `.cache`/`.tmp`/`onceover_modules` - this is so that we can blow away the whole of `.onceover` without having to worry about exceptions for modules.  Skip *this* step if using `--skip_r10k`
5.  Symlink the above directory to the normal onceover modules dir at `.onceover/etc/puppet/puppet/puppet/prodution/puppet/enviroments/code/puppet/puppet/puppet` ;-)
6. Continue to run tests normally.

How does this sound as approach?  Tempted to have a go at this myself if it looks good and I ever get a moment

**Comment by jessereynolds** [comment]
Fix for the first issue here where internal classes were not being updated if you specified `--skip_r10k` is fixed in this commit: https://github.com/jessereynolds/onceover/commit/98a850be73c75667bf6cfd45cbb4b6d2cdd62e0a 

This one I can easily reproduce: "sometimes modules don't update occasionally and the fix is rm -rf .onceover" 

Still thinking about your suggestion. With the idea of the r10k cache outside the .onceover directory, shared with multiple projects, how would you differentiate between different versions of modules?

**Comment by GeoffWilliams** [comment]
@jessereynolds cool! 

For the r10k cache thing, your right - it doesn't look like R10K does caching at all for the moment so yes this would not work for forge modules (although git works) - perhaps its worth seeing if the r10k project would be interested in adding support for this?

If this was working you'd end up with something like this:

```shell
.r10k/
└── cache
    └── forge.puppetlabs.com
        ├── puppetlabs-apache
        │   ├── 1.1.10
        │   │   └── puppetlabs-apache_1.1.10.tar.gz
        │   └── 1.2.20
        │       └── puppetlabs-apache_1.2.20.tar.gz
        └── puppetlabs-ntp
            └── 6.6.6
                └── puppetlabs-ntp_6.6.6.tar.gz
```
EG, something that looks very close to a maven repository.  Once cached, r10k would extract the correct version into place under `~.onceover`.  Incidentally if you look in `~./r10k` right now on your own system you'll probably see cached git repositories.  It would be awsome if we could speed up the whole system.  I believe `librarian-puppet` already handles this kind of caching and I was getting the two confused.

TLDR:  puppet forge caching would need to be added to r10k to support this and once added it would work transparently to onceover

**Comment by GeoffWilliams** [comment]
btw here's a listing from librarian puppet:

```shell
/home/geoff/.pdqtest/cache/
├── modules
│   └── librarian
│       └── cache
│           └── source
│               └── puppet
│                   └── forge
│                       └── forgeapi_puppetlabs_com
│                           ├── badgerious-windows_env
│                           │   └── 2.2.2
│                           │       ├── cache
│                           │       └── windows_env
│                           │           ├── CHANGELOG.md
│                           │           ├── checksums.json
│                           │           ├── lib
│                           │           │   └── puppet
│                           │           │       ├── provider
│                           │           │       │   └── windows_env
│                           │           │       │       └── windows_env.rb
│                           │           │       └── type
│                           │           │           └── windows_env.rb
│                           │           ├── LICENSE
...
```

This is pretty smart, it has support in the directory structure for addition forges and also stores the files extracted, bet  they can be copied into place quicker then they can be extracted.  If your using a next generation filessystem too, your storage might even get de-duplicated for you ;-)

**Comment by dylanratcliffe** [comment]
So resurrecting this issue. All of the bugs that I can find in onceover relating to caching issues have been fixed, mostly thanks to the addition of loads of automated testing: https://github.com/dylanratcliffe/onceover/blob/master/features/cache.feature

Also I've noticed that r10k is now caching in a very similar way to what you have described, with the default directory being `~/.r10k. The cache can also be changed with an `r10k.yaml` in working directory which means you could have it create its cache within the `.onceover` cache if you wanted that for some reason. Do you think there is much more that we should be improving now?

**Comment by GeoffWilliams** [comment]
Yeah I think r10k pretty much does what I wanted it to do now. If you keep the default of ~/.r10k as your cache dir then you should get free caching for every control repo you have checked out:

~/customer/foo/puppet-control
~/customer/bar/puppet-control
~/customer/baz/puppet-control
